components/nexus-repository/src/main/java/org/sonatype/nexus/repository/proxy/ProxyFacetSupport.java

```diff
@@ -262,36 +270,41 @@ private String extractETag(final Header etagHeader) {
   }
 
   /**
-   * For whatever component/asset is implied by the Context, return the date it was last verified up to date, or {@code
-   * null} if it isn't present.
-   */
-  @Nullable
-  protected abstract DateTime getCachedPayloadLastVerified(final Context context) throws IOException;
-
-  /**
    * For whatever component/asset
    */
-  protected abstract void indicateVerified(final Context context) throws IOException;
+  protected abstract void indicateVerified(final Context context, final CacheInfo cacheInfo) throws IOException;
 
   /**
    * Provide the URL of the content relative to the repository root.
    */
   protected abstract String getUrl(final @Nonnull Context context);
 
-  private boolean isStale(final Context context) throws IOException {
+  private boolean isStale(final Context context, final Content content) throws IOException {
+    if (content == null) {
+      // not in cache
+      return true;
+    }
+
     if (config.artifactMaxAge < 0) {
-      log.trace("Artifact max age checking disabled");
+      log.trace("Content max age checking disabled");
       return false;
     }
 
-    final DateTime lastUpdated = getCachedPayloadLastVerified(context);
-
-    if (lastUpdated == null) {
-      log.debug("Artifact last modified date unknown");
+    // FIXME: maybe require(CacheInfo.class), as cache content w/o CacheInfo should not exists in the first place
+    // FIXME: what happens with cache content once we go distributed?
+    final CacheInfo cacheInfo = content.getAttributes().get(CacheInfo.class);;
+    if (cacheInfo == null) {
+      log.debug("Content cache info unknown");
       return true;
     }
-
-    final DateTime earliestFreshDate = new DateTime().minusMinutes(config.artifactMaxAge);
-    return lastUpdated.isBefore(earliestFreshDate);
+    if (cacheToken != null && !cacheToken.equals(cacheInfo.getCacheToken())) {
+      log.debug("Content expired (cacheToken)");
```

Comment:
OK, anything else we can log to help tie the expired cache to the resource? Just thinking it might make it easier to trace things when there are interleaved requests.